File: Sales-Force-Automation/sfa-sales-performance-report.js
const d3 = require('d3');
const fs = require('fs');
const sqlite3 = require('sqlite3');

// Connect to the database
const db = new sqlite3.Database('sales.db');

// Retrieve sales performance data from the database
let salesPerformanceData = [];
db.all("SELECT s.name, SUM(sp.amount) as total_sales, sq.amount as sales_quota FROM salespeople s JOIN sales_performance sp ON s.salesperson_id = sp.salesperson_id JOIN sales_quotas sq ON s.salesperson_id = sq.salesperson_id GROUP BY s.salesperson_id", (err, rows) => {
    if (err) {
        throw err;
    }
    salesPerformanceData = rows;
});

// Create an HTML template with a table to display the sales performance data
const html = `
<html>
  <head>
    <style>
      table {
        font-family: arial, sans-serif;
        border-collapse: collapse;
        width: 100%;
      }

      td, th {
        border: 1px solid #dddddd;
        text-align: left;
        padding: 8px;
      }

      th {
        background-color: #dddddd;
      }
    </style>
  </head>
  <body>
    <h1>Sales Performance Report</h1>
    <table>
      <tr>
        <th>Name</th>
        <th>Total Sales</th>
        <th>Sales Quota</th>
        <th>Achieved Quota</th>
      </tr>
      ${d3.select('body')
        .selectAll('tr')
        .data(salesPerformanceData)
        .enter()
        .append('tr')
        .html(d => {
          return `
            <td>${d.name}</td>
            <td>${d.total_sales}</td>
            <td>${d.sales_quota}</td>
            <td>${d.achieved_quota ? 'Yes' : 'No'}</td>
          `;
        })}
    </table>
  </body>
</html>
`;

// Write the HTML template to an HTML file
fs.writeFileSync('sales_performance_report.html', html);

File: Sales-Force-Automation/sfa-sales-performance-report.py
import mysql.connector
import csv
from tabular import tabular

# Connect to the database
mydb = mysql.connector.connect(
  host="hostname",
  user="username",
  password="password",
  database="databasename"
)

mycursor = mydb.cursor()

# Retrieve sales performance data from the database
mycursor.execute("SELECT s.name, SUM(sp.amount) as total_sales, sq.amount as sales_quota FROM salespeople s JOIN sales_performance sp ON s.salesperson_id = sp.salesperson_id JOIN sales_quotas sq ON s.salesperson_id = sq.salesperson_id GROUP BY s.salesperson_id")
sales_performance = mycursor.fetchall()

# Create a list of dictionaries with the sales performance data
sales_performance_data = []
for row in sales_performance:
    sales_performance_data.append({
        'name': row[0],
        'total_sales': row[1],
        'sales_quota': row[2],
        'achieved_quota': row[1] >= row[2]
    })

# Sort the sales performance data by total sales
sales_performance_data = sorted(sales_performance_data, key=lambda x: x['total_sales'], reverse=True)

# Create a CSV file with the sales performance data
with open('sales_performance.csv', mode='w') as csv_file:
    fieldnames = ['name', 'total_sales', 'sales_quota', 'achieved_quota']
    writer = csv.DictWriter(csv_file, fieldnames=fieldnames)
    writer.writeheader()
    for data in sales_performance_data:
        writer.writerow(data)

# Create a plain text file with the sales performance data using tabular library
with open('sales_performance.txt', mode='w') as txt_file:
    t = tabular.Tabular()
    t.add_column('Name', [x['name'] for x in sales_performance_data])
    t.add_column('Total Sales', [x['total_sales'] for x in sales_performance_data])
    t.add_column('Sales Quota', [x['sales_quota'] for x in sales_performance_data])
    t.add_column('Achieved Quota', [x['achieved_quota'] for x in sales_performance_data])
    txt_file.write(t.render())

# Close the cursor and the database connection
mycursor.close()
mydb.close()
